src/types/Slot0.sol

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+/*
+    Slot0 is a packed version of solidity structure:
+
+    // the current price
+    uint160 sqrtPriceX96;
+    // the current tick
+    int24 tick;
+    // protocol fee, expressed in hundredths of a bip
+    // upper 12 bits are for 1->0, and the lower 12 are for 0->1
+    // the maximum is 1000 - meaning the maximum protocol fee is 0.1%
+    // the protocolFee is taken from the input first, then the lpFee is taken from the remaining input
+    uint24 protocolFee;
+    // used for the lp fee, either static at initialize or dynamic via hook
+    uint24 lpFee;
+*/
+
+type Slot0 is bytes32;
+
+using Slot0Library for Slot0 global;
+
+library Slot0Library {
+    uint160 private constant UINT160_MASK = 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+    uint24 private constant INT24_MASK = 0xFFFFFF;
+    uint24 private constant UINT24_MASK = 0xFFFFFF;
+
+    uint8 private constant TICK_OFFSET = 160;
+    uint8 private constant PROTOCOL_FEE_OFFSET = 184;
+    uint8 private constant LP_FEE_OFFSET = 208;
+
+    // #### GETTERS ####
+    function sqrtPriceX96(Slot0 _packed) internal pure returns (uint160 _sqrtPriceX96) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _sqrtPriceX96 := and(UINT160_MASK, _packed)
+        }
+    }
+
+    function tick(Slot0 _packed) internal pure returns (int24 _tick) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _tick := signextend(2, shr(TICK_OFFSET, _packed))
+        }
+    }
+
+    function protocolFee(Slot0 _packed) internal pure returns (uint24 _protocolFee) {
```

Comment:
I wonder now that we are writing getters manually if we can just get rid of the getters for 0 to 1 and 1 to 0 in ProtocolFeeLibrary and just do it directly in this library? Maybe would save some overhead - I still like having a ProtcocolFeeLibrary but worth playing around if we can remove getZeroForOneFee / getOneForZeroFee

Reply:
Sounds reasonable! Then we can remove `protocolFee` completely and treat it as two packed `uint12` inside of a structure.

In this case, it is logical to change the public getter for slot0 too

Reply:
I tried to implement `protocolFeeZeroForOne`/`protocolFeeOneForZero` as `uint12` in `Slot0` and do not see any gains in snapshots, looks like the optimizer itself substitutes and optimizes getters.
Probably such a change only makes sense to simplify the code, but imo the current version is already quite good.

Reply:
Simple getters without side effects should be automatically inlined by the compiler.